### PR TITLE
Added [[nodiscard]] on dl_error_t and handled the fallout

### PR DESCRIPTION
**Pull request by Tisten**
If the code is compiled with C++17

### CODE REVIEW

File: src/dl_txt_unpack.cpp [hunk 1]
```diff
@@ -623,20 +630,23 @@ static void dl_txt_unpack_write_member_subdata(dl_ctx_t dl_ctx, dl_txt_unpack_ct
...
+																  dl_internal_find_type( dl_ctx, member->type_id ) );
 				break;
 				case DL_TYPE_STORAGE_STRUCT:
 				{
 					const dl_type_desc* subtype = dl_internal_find_type( dl_ctx, member->type_id );
 					if( subtype->flags & DL_TYPE_FLAG_HAS_SUBDATA )
 					{
 						for( uint32_t i = 0; i < member->inline_array_cnt(); ++i )
-							dl_txt_unpack_write_subdata( dl_ctx, unpack_ctx, writer, subtype, member_data + i * subtype->size[DL_PTR_SIZE_HOST] );
+				        {
```
Comment (wc-duck) [inline src/dl_txt_unpack.cpp:646]: Indent

File: src/dl_txt_unpack.cpp [hunk 2]
```diff
@@ -659,7 +669,10 @@ static void dl_txt_unpack_write_member_subdata(dl_ctx_t dl_ctx, dl_txt_unpack_ct
 						uint32_t  array_count  = *(uint32_t*)(member_data + sizeof(uintptr_t));
 						const uint8_t* array = unpack_ctx->packed_instance + array_offset;
 						for( uint32_t i = 0; i < array_count; ++i )
-							dl_txt_unpack_write_subdata( dl_ctx, unpack_ctx, writer, subtype, array + i * subtype->size[DL_PTR_SIZE_HOST] );
+				        {
```
Comment (wc-duck) [inline src/dl_txt_unpack.cpp:672]: One more

File: src/dl_typelib_write_c_header.cpp [hunk 3]
```diff
@@ -222,43 +223,50 @@ static void dl_context_write_c_header_end( dl_binary_writer* writer, const char*
...
-	dl_reflect_context_info( ctx, &ctx_info );
+	dl_error_t err = dl_reflect_context_info( ctx, &ctx_info );
+	if( err != DL_ERROR_OK )
+		return err;
 
 	if(ctx_info.num_enums != 0)
 	{
-		dl_enum_info_t* enum_infos = (dl_enum_info_t*)malloc( ctx_info.num_enums * sizeof( dl_enum_info_t ) );
-		dl_reflect_loaded_enums(ctx, enum_infos, ctx_info.num_enums);
+		unique_malloc_ptr<dl_enum_info_t[]> enum_infos( (dl_enum_info_t*)malloc( ctx_info.num_enums * sizeof(dl_enum_info_t) ), free );
```
Comment (wc-duck) [inline src/dl_typelib_write_c_header.cpp:242]: Did not see this before? If you want automatic cleanup I would prefer a DL_DEFERED-macro

Comment (wc-duck) [reply src/dl_typelib_write_c_header.cpp:242]: It is also not prefixed with dl_

File: src/dl_typelib_write_c_header.cpp [hunk 4]
```diff
@@ -492,64 +508,67 @@ static void dl_context_write_type( dl_ctx_t ctx, dl_type_storage_t storage, dl_t
...
+		case DL_TYPE_STORAGE_INT16:    dl_binary_writer_write_string_fmt(writer, "int16_t"); return DL_ERROR_OK;
+		case DL_TYPE_STORAGE_UINT16:   dl_binary_writer_write_string_fmt(writer, "uint16_t"); return DL_ERROR_OK;
+		case DL_TYPE_STORAGE_INT32:    dl_binary_writer_write_string_fmt(writer, "int32_t"); return DL_ERROR_OK;
+		case DL_TYPE_STORAGE_UINT32:   dl_binary_writer_write_string_fmt(writer, "uint32_t"); return DL_ERROR_OK;
+		case DL_TYPE_STORAGE_INT64:    dl_binary_writer_write_string_fmt(writer, "DL_ALIGN(8) int64_t"); return DL_ERROR_OK;
+		case DL_TYPE_STORAGE_UINT64:   dl_binary_writer_write_string_fmt(writer, "DL_ALIGN(8) uint64_t"); return DL_ERROR_OK;
+		case DL_TYPE_STORAGE_FP32:     dl_binary_writer_write_string_fmt(writer, "float"); return DL_ERROR_OK;
+		case DL_TYPE_STORAGE_FP64:     dl_binary_writer_write_string_fmt(writer, "DL_ALIGN(8) double"); return DL_ERROR_OK;
+		case DL_TYPE_STORAGE_STR:    dl_binary_writer_write_string_fmt(writer, "const char*"); return DL_ERROR_OK;
+	    case DL_TYPE_STORAGE_PTR:
```
Comment (wc-duck) [inline src/dl_typelib_write_c_header.cpp:536]: What happened here?

File: include/dl/dl.h [hunk 5]
```diff
@@ -60,6 +60,9 @@ static const int DL_INLINE_ARRAY_LENGTH_MAX = 0xFFFF;
 	DL_ERROR_INTERNAL_ERROR                                - Internal error, contact dev!
 */
 typedef enum
+#if( _MSVC_LANG >= 201703L ) || ( __cplusplus >= 201703L )
```
[the remark this answers is not in the document]
Comment (Tisten) [reply include/dl/dl.h]: Done

File: src/dl_txt_pack.cpp [hunk 6]
```diff
@@ -865,7 +869,10 @@ static void dl_txt_pack_member( dl_ctx_t dl_ctx, dl_txt_pack_ctx* packctx, size_
 			}
 
 			if(array_length > 0)
-				dl_txt_pack_eat_and_write_array( dl_ctx, packctx, member, array_length );
+		    {
```
[the remark this answers is not in the document]
Comment (Tisten) [reply src/dl_txt_pack.cpp:872]: Would you be opposed to creating a .clang-format file which I can autoformat these things? :)

My current 6 open PRs are full of places where they confict each other, plus that I have a few other WIP tasks going on. Getting the formatting right manually is terribly time-consuming.

File: src/dl_txt_unpack.cpp (hunk 2 above)
Comment (Tisten) [reply src/dl_txt_unpack.cpp:672]: This is my Visual Studio switching tabs to spaces in some places, and I don't know why yet. And with tabs set to 4 spaces you don't notice these. :(

File: src/dl_typelib_write_c_header.cpp (hunk 4 above)
Comment (Tisten) [reply src/dl_typelib_write_c_header.cpp:536]: Visual Studio wanted spaces instead of tabs...

File: src/dl_typelib_write_c_header.cpp (hunk 3 above)
Comment (Tisten) [reply src/dl_typelib_write_c_header.cpp:242]: Do you mean a special "run this lambda in the destructor" define?

File: src/dl_txt_pack.cpp (hunk 6 above)
[the remark this answers is not in the document]
Comment (wc-duck) [reply src/dl_txt_pack.cpp]: I should probably do that yes...

File: src/dl_typelib_write_c_header.cpp (hunk 3 above)
Comment (wc-duck) [reply src/dl_typelib_write_c_header.cpp]: Something like this yes https://www.gingerbill.org/article/2015/08/19/defer-in-cpp/

**Review by wc-duck — APPROVED** [review]

**Pull request merged**